src/wasm/wasm-validator.cpp

```diff
@@ -2845,23 +2852,66 @@ static void validateTables(Module& module, ValidationInfo& info) {
     }
   }
 
+  for (auto& table : module.tables) {
+    info.shouldBeTrue(table->initial <= table->max,
+                      "table",
+                      "size minimum must not be greater than maximum");
+    info.shouldBeTrue(
+      table->type.isNullable(),
+      "table",
+      "Non-nullable reference types are not yet supported for tables");
+    if (!module.features.hasGC()) {
+      info.shouldBeTrue(
+        table->type.isFunction() || table->type == Type::externref,
+        "table",
+        "Only function reference types or externref are valid for table type.");
+    }
+    if (!module.features.hasTypedFunctionReferences()) {
+      info.shouldBeTrue(table->type == Type::funcref ||
+                          table->type == Type::externref,
+                        "table",
+                        "Only funcref and externref are valid for table type.");
+    }
+  }
+
   for (auto& segment : module.elementSegments) {
+    info.shouldBeTrue(segment->type != Type::externref,
+                      "elem",
+                      "element segment type cannot be externref");
```

Comment:
This is surprising!

Reply:
Again, this is only my understanding of the spec but haven't seen the spec text mention it anywhere, nor any spec tests that expects this to be invalid. But, here's my reasoning:

Element segments are only used to initialize tables and cannot be imported from outside. Given the fact that they cannot include function call expressions, there's no way for them to hold `externref` values. The only way for a wasm module to change items on an `externref` table is via the table instructions like `table.set` and `table.copy` that are allowed at runtime.

Reply:
This is now changed to only allow function types for element segments. I added a comment block explaining the reason.